### PR TITLE
Feature/elipsoids filter

### PR DESCRIPTION
**Pull request by gawela**
This adds ElipsoidsFilter and 3D Gestalt Filter Implementations.

ElipsoidsFilter:
- calculates elipsoids around keypoints based on the neighbourhood

GestaltFilter:
- calculates the 3D Gestalt features around keypoints, see:
Bosse, Michael, and Robert Zlot. "Place recognition using keypoint voting in large 3D lidar datasets." Robotics and Automation (ICRA), 2013 IEEE International Conference on. IEEE, 2013.

### CODE REVIEW

**Comment by pomerlef** [comment]
One of the unit test failed:

```
/var/lib/jenkins/jobs/libpointmatcher/workspace/utest/utest.cpp:163
Expected: icp.loadFromYaml(ifs) doesn't throw an exception.
  Actual: it throws.
This error was caused by the test file:
   ../examples/data/default-identity.yaml
```

File: pointmatcher/DataPointsFiltersImpl.cpp [hunk 1]
```diff
@@ -595,7 +596,47 @@ typename PointMatcher<T>::Vector DataPointsFiltersImpl<T>::SurfaceNormalDataPoin
 		}
 	}
 
-	return eigenVe.col(smallestId);
+  return eigenVe.col(smallestId);
+}
+
+template<typename T>
+Eigen::Matrix<T,3,1> DataPointsFiltersImpl<T>::SurfaceNormalDataPointsFilter::sortEigenValues(const Vector eigenVa) {
```
Comment (pomerlef) [inline pointmatcher/DataPointsFiltersImpl.cpp]: Is there a way to generalize this function for any dimensions? Or at least 2D and 3D?

File: pointmatcher/DataPointsFiltersImpl.cpp [hunk 2]
```diff
@@ -940,6 +950,994 @@ void DataPointsFiltersImpl<T>::SamplingSurfaceNormalDataPointsFilter::fuseRange(
...
+      // throw out surfel if it does not meet planarity criteria
+      if (planarity < minPlanarity)
+      {
+        data.unfitPointsCount += colCount;
+        return;
+      }
+    }
+  }
+
+  // keep the indices of each ellipsoid?
```
Comment (rdube) [inline pointmatcher/DataPointsFiltersImpl.cpp]: Can we address this comment?

File: pointmatcher/DataPointsFiltersImpl.cpp [hunk 3]
```diff
@@ -940,6 +950,994 @@ void DataPointsFiltersImpl<T>::SamplingSurfaceNormalDataPointsFilter::fuseRange(
...
+      pointIds(i) = data.indices[first+i];
+      points.col(i) = data.features.block(0,data.indices[first+i],2, 1);
+    }
+  }
+
+  Vector normal;
+  if(keepNormals)
+    normal = SurfaceNormalDataPointsFilter::computeNormal(eigenVa, eigenVe);
+
+  T densitie = 0;
```
Comment (rdube) [inline pointmatcher/DataPointsFiltersImpl.cpp]: Maybe "density" ?

File: pointmatcher/DataPointsFiltersImpl.cpp [hunk 4]
```diff
@@ -940,6 +950,994 @@ void DataPointsFiltersImpl<T>::SamplingSurfaceNormalDataPointsFilter::fuseRange(
...
+        if(keepMeans)
+          data.means->col(k) = mean;
+        // a 3d vecetor of shape parameters: planarity (P), cylindricality (C), sphericality (S)
+        if(keepShapes) {
+          Eigen::Matrix<T, 3, 3> shapeMat;
+          (shapeMat << 0, 2, -2, 1, -1, 0, 0, 0, 3);
+          Eigen::Matrix<T, 3, 1> vals;
+          (vals << eigenVa(0),eigenVa(1),eigenVa(2));
+          vals = vals/eigenVa.sum();
+          data.shapes->col(k) = shapeMat * vals;//eigenVa;
```
Comment (rdube) [inline pointmatcher/DataPointsFiltersImpl.cpp]: remove old code

File: pointmatcher/DataPointsFiltersImpl.cpp [hunk 5]
```diff
@@ -940,6 +950,994 @@ void DataPointsFiltersImpl<T>::SamplingSurfaceNormalDataPointsFilter::fuseRange(
...
+  // i, j, k are the component indices
+  // nx, ny number of divisions in x and y components
+  // idx = i + j * nx + k * nx * ny
+  int numPoints = last - first;
+  std::vector<unsigned int> indices(numPoints);
+
+  // vector to hold the first point in a voxel
+  // this point will be ovewritten in the input cloud with
+  // the output value
+
```
Comment (rdube) [inline pointmatcher/DataPointsFiltersImpl.cpp]: remove empty line

File: pointmatcher/DataPointsFiltersImpl.cpp [hunk 6]
```diff
@@ -940,6 +950,994 @@ void DataPointsFiltersImpl<T>::SamplingSurfaceNormalDataPointsFilter::fuseRange(
...
+      // Keep points with their descriptors
+      int k = pointsToKeep[i];
+      // Mark the indices which will be part of the final data
+      data.indicesToKeep.push_back(k);
+    }
+  }
+}
+
+template<typename T>
+void DataPointsFiltersImpl<T>::GestaltDataPointsFilter::fuseRanger(BuildData& data, DataPoints& input, const int first, const int last) const
```
Comment (rdube) [inline pointmatcher/DataPointsFiltersImpl.cpp]: What is the difference between fuseRange() and fuseRanger() ?

File: pointmatcher/DataPointsFiltersImpl.cpp [hunk 7]
```diff
@@ -940,6 +950,994 @@ void DataPointsFiltersImpl<T>::SamplingSurfaceNormalDataPointsFilter::fuseRange(
...
+  typedef typename MatchersImpl<T>::KDTreeMatcher KDTreeMatcher;
+  typedef typename PointMatcher<T>::Matches Matches;
+
+  const int featDim(data.features.rows());
+  std::vector<int> indicesToKeepStrict;
+  for (int i = 0; i< data.indicesToKeep.size(); ++i) {
+    Eigen::Matrix<T,3,1> keyPoint;
+    keyPoint = input.features.col(data.indicesToKeep[i]);
+
+    // make a nearest neighbour hack around keyPoint with box:
```
Comment (rdube) [inline pointmatcher/DataPointsFiltersImpl.cpp]: Can the comment be more descriptive?

File: pointmatcher/DataPointsFiltersImpl.cpp [hunk 8]
```diff
@@ -940,6 +950,994 @@ void DataPointsFiltersImpl<T>::SamplingSurfaceNormalDataPointsFilter::fuseRange(
...
+      heights = data.warpedXYZ->row(2);
+
+      // sort points into Gestalt bins
+      T angularBinWidth = M_PI/4;
+      T radialBinWidth = radius/4;
+      Matrix indices(2, colCount);
+      gestaltMeans = Matrix::Zero(4, 8);
+      gestaltVariances = Matrix::Zero(4, 8);
+      numOfValues = Matrix::Zero(4, 8);
+      // sort the points into right bins
```
Comment (rdube) [inline pointmatcher/DataPointsFiltersImpl.cpp]: repetition of comment 7 lines above

File: pointmatcher/DataPointsFiltersImpl.cpp [hunk 9]
```diff
@@ -940,6 +950,994 @@ void DataPointsFiltersImpl<T>::SamplingSurfaceNormalDataPointsFilter::fuseRange(
...
+
+      for (int it=0; it < colCount; ++it) {
+        indices(0,it) = floor(radii(it)/radialBinWidth);//floor(angles(it)/angularBinWidth);
```
Comment (rdube) [inline pointmatcher/DataPointsFiltersImpl.cpp]: remove old code

File: utest/ui/DataFilters.cpp [hunk 10]
```diff
@@ -256,6 +256,50 @@ TEST_F(DataFilterTest, SamplingSurfaceNormalDataPointsFilter)
 
 }
 
+TEST_F(DataFilterTest, ElipsoidsDataPointsFilter)
+{
+	// This filter create descriptor AND subsample
```
Comment (rdube) [inline utest/ui/DataFilters.cpp]: create(s) and subsample(s) ?

File: utest/ui/DataFilters.cpp [hunk 11]
```diff
@@ -256,6 +256,50 @@ TEST_F(DataFilterTest, SamplingSurfaceNormalDataPointsFilter)
...
+	params["keepIndices"] = "1";
+
+	addFilter("ElipsoidsDataPointsFilter", params);
+	validate2dTransformation();
+	validate3dTransformation();
+}
+
+TEST_F(DataFilterTest, GestaltDataPointsFilter)
+{
+	// This filter create descriptor AND subsample
```
Comment (rdube) [inline utest/ui/DataFilters.cpp]: same as above

**Comment by rdube** [comment]
scolgtm! Good work!

**Comment by pomerlef** [comment]
@gawela I'm seeing bunch of warning for unused variables since that merge. Could you have a look into that:

Example:
```
[...]
/Users/francoispomerleau/Research/Code/libpointmatcher/pointmatcher/DataPointsFiltersImpl.cpp:1934:47: note: in instantiation of member function
      'DataPointsFiltersImpl<float>::GestaltDataPointsFilter::inPlaceFilter' requested here
template struct DataPointsFiltersImpl<float>::GestaltDataPointsFilter;
                                              ^
[...]
```

**Comment by gawela** [comment]
@pomerlef Addressed the warnings in #127 
Sorry!

**Pull request merged**